File: python-to-go-upskill/Day1a/code/string.py
"""
Reverse a string in Python using slicing.

This example shows how to get the reverse of a string using the slice
syntax `s[start:stop:step]` with a negative step value.

How it works:
- `s[::-1]` means:
  - start from the end of the string
  - go backwards one character at a time
  - stop when the beginning is reached

Time complexity:
- O(n), where n is the length of the string. Python must visit each character once.

Space complexity:
- O(n), because reversing creates a new string of the same length.

Example:
- Input  : "upskill"
- Output : "llikspu"
"""

s = "upskill"
reversed_s = s[::-1]
print(reversed_s)

# Input  : "upskill"
# Output : "llikspu"

File: python-to-go-upskill/Day1a/code/string.go
package main

import "fmt"

/*
reverseString reverses the characters in a string.

This program is written for fresh learners.
The reverseString function converts the input string to a slice of runes
so it handles multi-byte characters correctly. It then swaps characters
from the outside in until the middle is reached.

Example:
- Input  : "upskill"
- Output : "llikspu"
*/
func reverseString(s string) string {
	r := []rune(s)
	for i, j := 0, len(r)-1; i < j; i, j = i+1, j-1 {
		r[i], r[j] = r[j], r[i]
	}
	return string(r)
}

func main() {
	result := reverseString("upskill")
	fmt.Println(result)

	// Complexity explanation:
	// Time complexity: O(n)
	//   We visit each character in the string once when swapping.
	// Space complexity: O(n)
	//   We create a new slice of runes and a new string of the same length.
}
